python/perf-kernels/flash-attention.py

```diff
@@ -127,6 +164,19 @@ def _attn_fwd_inner(
         m_ij = tl.maximum(m_i, tl.max(qk, 1))
         qk = qk - m_ij[:, None]
         p = tl.math.exp2(qk)
+        # CAVEAT: Must update l_ij before applying dropout
+        l_ij = tl.sum(p, 1)
+        # Note about the conflicts of Flash attention algorithm and PyTorch's CUDA implementation
+        # PyTorch needs to return softmax(qk) (dropout mask encoded in sign bits)
+        # While Flash attention paper computer the dropout AFTER exp2(qk- m_ij)
+        if ENABLE_DROPOUT:
+            philox_offset = batch_philox_offset + start_m * BLOCK_M * seqlen_k + start_n
+            keep = dropout_mask(philox_seed, philox_offset, dropout_p, BLOCK_M, BLOCK_N, seqlen_k)
+            if RETURN_ENCODED_SOFTMAX:
+                tl.store(encoded_softmax_block_ptr, tl.where(keep, p, -p).to(encoded_softmax_block_ptr.type.element_ty))
+            p = tl.where(keep, p, 0.0)
+        elif RETURN_ENCODED_SOFTMAX:
+            tl.store(encoded_softmax_block_ptr, p.to(encoded_softmax_block_ptr.type.element_ty))
```

Comment:
Does this need to be boundary_checked if seqlen_k or seqlen_q isn't a multiple of BLOCK_N/M?

Reply:
Yes, we'll add this later.
`encoded_softmax` was not well designed in this kernel since it is fundamentally a debugging feature.
Suppose unaligned store is supported by the GPU, we won't even have plan to pad this tensor.

Reply:
I'd add at least a tl.static_assert to check for this (more preferable), or a comment and a TODO.

Debugging issues here at a future data because we accidentally enabled encoded_softmax with a non-multiple of block size sequence length will be painful.

Reply:
Noted, I'll add an assert.